core/src/main/java/org/apache/iceberg/data/GenericRecord.java

```diff
@@ -115,15 +120,27 @@ public Object get(int pos) {
   }
 
   @Override
+  @SuppressWarnings("unchecked")
   public <T> T get(int pos, Class<T> javaClass) {
     Object value = get(pos);
     if (value == null || javaClass.isInstance(value)) {
       return javaClass.cast(value);
+    }
+
+    if (value instanceof LocalDateTime) {
```

Comment:
This looks suspicious. The caller is responsible for translating between the representations used by object models. It looks like what's happening here is that generics use a different one (`LocalDateTime`) than internals (long micros). Usually we wrap the record to convert, since this changing this everywhere is going to break a ton of things that use the generic object model.